### PR TITLE
libm/newlib: add newlib/libm support 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

libm/newlib: add newlib/libm support 

Picolibc used by llvm armclang on v15.0 merges libm into libc: https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/issues/158

Since nuttx provides libc implementation, which cannot reference toolchain libc, so in this PR, we added support for newlib/libm, which will help developers using llvm armclang to still use the math library with related architecture optimization support

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

**Comment by anchao** [comment]
please ignore the style check error : Mixed case identifier found

```
../nuttx/tools/checkpatch.sh -u -m -g b864f376135c7055d982e95ed6d862ef8f07330e..HEAD
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:59:7: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:[13](https://github.com/apache/nuttx/actions/runs/4423582588/jobs/7756456883#step:3:14)6:2: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:[14](https://github.com/apache/nuttx/actions/runs/4423582588/jobs/7756456883#step:3:15)0:20: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:445:9: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:447:9: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:448:9: error: Mixed case identifier found
Error: Process completed with exit code 1.
```

File: libs/libm/newlib/include/_newlib_version.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef _NEWLIB_VERSION_H__
+#define _NEWLIB_VERSION_H__ 1
```
Comment (pkarashchenko) [inline libs/libm/newlib/include/_newlib_version.h:22]:
```suggestion
#ifndef __LIBS_LIBM_NEWLIB_INCLUDE__NEWLIB_VERSION_H
#define __LIBS_LIBM_NEWLIB_INCLUDE__NEWLIB_VERSION_H
```

File: libs/libm/newlib/Kconfig [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#
+
+# These are library functions that may be overridden by architecture-
+# specific implementations.  Not all architectures support implementations
+# for every library function.
+
+# Default settings for C library functions that may be replaced with
+# architecture-specific versions.
+
+config LIBM_NEWLIB_HWFP
```
[the remark this answers is not in the document]
Comment (anchao) [reply libs/libm/newlib/Kconfig:13]: It seems that the feature of fp-hw is disabled by newlib, but the code is still there, I think we can remove this kconfig:
https://github.com/mirror/newlib-cygwin/commit/6555743a4bba74913f8df26482bbd9ed4f8adddf

File: libs/libm/newlib/include/_newlib_version.h (hunk 1 above)
Comment (anchao) [reply libs/libm/newlib/include/_newlib_version.h:22]: Done

File: libs/libm/newlib/Kconfig (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchao) [reply libs/libm/newlib/Kconfig]: Done, @pkarashchenko please review again

**Review by acassis — APPROVED** [review]

File: libs/libm/Makefile [hunk 3]
```diff
@@ -23,6 +23,9 @@ include $(TOPDIR)/Make.defs
 ifeq ($(CONFIG_LIBM),y)
 include libm/Make.defs
 endif
+ifeq ($(CONFIG_LIBM_NEWLIB),y)
```
Comment (xiaoxiang781216) [inline libs/libm/Makefile:26]: else ifeq

File: libs/libm/newlib/include/math.h [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <stdint.h>
+
+#include <sys/features.h>
+#include <machine/ieeefp.h>
+#include <_newlib_version.h>
```
Comment (xiaoxiang781216) [inline libs/libm/newlib/include/math.h:33]: let's remove the non standard header file: sys/feature.h _newlib_version.h and sys/cdefs.h

File: libs/libm/newlib/include/math.h [hunk 5]
```diff
@@ -0,0 +1,775 @@
+/****************************************************************************
+ * libs/libm/newlib/include/math.h
```
Comment (xiaoxiang781216) [inline libs/libm/newlib/include/math.h:2]: need add complex.h

File: libs/libm/newlib/0001-newlib-libm-fix-NAN-redefined-if-enable-enable-newli.patch [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ newlib/libm/common/math_config.h | 16 ++++++++++++----
+ newlib/libm/mathfp/zmath.h       |  5 ++++-
+ 2 files changed, 16 insertions(+), 5 deletions(-)
+
+diff --git a/newlib/libm/common/math_config.h newlib/newlib/newlib/libm/common/math_config.h
+index 0f78b5c09..9116c754f 100644
+--- a/newlib/libm/common/math_config.h
++++ newlib/newlib/newlib/libm/common/math_config.h
+@@ -253,12 +253,20 @@ eval_as_double (double x)
+
```
Comment (pkarashchenko) [inline libs/libm/newlib/0001-newlib-libm-fix-NAN-redefined-if-enable-enable-newli.patch:28]: should we add `#include <nuttx/compiler.h>`?

File: libs/libm/newlib/include/_newlib_version.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+#define __LIBS_LIBM_NEWLIB_INCLUDE__NEWLIB_VERSION_H
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define _NEWLIB_VERSION "4.3.0"
+#define __NEWLIB__ 4
+#define __NEWLIB_MINOR__ 3
+#define __NEWLIB_PATCHLEVEL__ 0
```
Comment (pkarashchenko) [inline libs/libm/newlib/include/_newlib_version.h:31]: Maybe we can `-D` this from `libs/libm/newlib/Make.defs`? Or this file is needed inside the newly code?

File: libs/libm/newlib/include/ieeefp.h [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+#ifdef __cplusplus
+#define EXTERN  extern "C"
+extern "C"
+{
+#else
+#define EXTERN  extern
+#endif
+
+#ifdef __IEEE_BIG_ENDIAN
```
Comment (pkarashchenko) [inline libs/libm/newlib/include/ieeefp.h:45]: Where `__IEEE_BIG_ENDIAN` is defined?

**Comment by PasunuriSrinidhi** [comment]
> please ignore the style check error : Mixed case identifier found
> 
> ```
> ../nuttx/tools/checkpatch.sh -u -m -g b864f376135c7055d982e95ed6d862ef8f07330e..HEAD
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:59:7: error: Mixed case identifier found
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:[13](https://github.com/apache/nuttx/actions/runs/4423582588/jobs/7756456883#step:3:14)6:2: error: Mixed case identifier found
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math.h:[14](https://github.com/apache/nuttx/actions/runs/4423582588/jobs/7756456883#step:3:15)0:20: error: Mixed case identifier found
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:445:9: error: Mixed case identifier found
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:447:9: error: Mixed case identifier found
> Error: /home/runner/work/nuttx/nuttx/nuttx/libs/libm/newlib/include/math_config.h:448:9: error: Mixed case identifier found
> Error: Process completed with exit code 1.
> ```

To fix this error, you should update the code to use consistent capitalization for identifiers. This means using all upper-case letters for constants, and using lower-case letters for variables and functions. Once you've made the necessary changes, you can rerun the checkpatch.sh script to ensure that the code is now compliant with the NuttX coding style.

**Comment by xiaoxiang781216** [comment]
close, finish here: https://github.com/apache/nuttx/pull/9061

**Pull request closed without merging**